### PR TITLE
feat: add enableSnappy flag to import snapshot pipeline and select th…

### PR DESCRIPTION
**Pull request by TracyCuiCan**
…e right container image based on beam version if flag is set.

Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/java-bigtable-hbase/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> ☕️

If you write sample code, please follow the [samples format](
https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/SAMPLE_FORMAT.md).

### CODE REVIEW

File: bigtable-dataflow-parent/bigtable-beam-import/src/test/integration-test/data/.hbase-snapshot/snappy-compressed-snapshot/.snapshotinfo [hunk 1]
```diff
@@ -0,0 +1,2 @@
+
+demo-snapshotdemo-1����0 (
```
Comment (vermas2012) [inline bigtable-dataflow-parent/bigtable-beam-import/src/test/integration-test/data/.hbase-snapshot/snappy-compressed-snapshot/.snapshotinfo:2]: this snapshot name is not the same as test file? does this work?

File: bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/hbasesnapshots/ImportJobFromHbaseSnapshot.java [hunk 2]
```diff
@@ -87,6 +97,18 @@ public static void main(String[] args) throws Exception {
...
 
+    if(opts.getEnableSnappy() != null && opts.getEnableSnappy()) {
+        DataflowPipelineOptions dataFlowOpts = opts.as(DataflowPipelineOptions.class);
+        dataFlowOpts.setSdkContainerImage(CONTAINER_IMAGE_PATH_PREFIX + ReleaseInfo.getReleaseInfo().getVersion());
+        List<String> expOpts = dataFlowOpts.getExperiments();
+        if (expOpts != null) {
+          expOpts.add("use_runner_v2");
+          dataFlowOpts.setExperiments(expOpts);
+        } else {
+          dataFlowOpts.setExperiments(Arrays.asList("use_runner_v2"));
```
Comment (vermas2012) [inline bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/hbasesnapshots/ImportJobFromHbaseSnapshot.java]: How do we execute both the code paths from test?

File: bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/hbasesnapshots/ImportJobFromHbaseSnapshot.java [hunk 3]
```diff
@@ -87,6 +97,18 @@ public static void main(String[] args) throws Exception {
     ImportOptions opts =
         PipelineOptionsFactory.fromArgs(args).withValidation().as(ImportOptions.class);
 
+    if(opts.getEnableSnappy() != null && opts.getEnableSnappy()) {
+        DataflowPipelineOptions dataFlowOpts = opts.as(DataflowPipelineOptions.class);
+        dataFlowOpts.setSdkContainerImage(CONTAINER_IMAGE_PATH_PREFIX + ReleaseInfo.getReleaseInfo().getVersion());
```
Comment (vermas2012) [inline bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/hbasesnapshots/ImportJobFromHbaseSnapshot.java]: Is there a way to fail fast here? Can we check if the image exists at this point?

File: clirr-ignored-differences.xml [hunk 4]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!-- see http://www.mojohaus.org/clirr-maven-plugin/examples/ignored-differences.html -->
+<differences>
+    <difference>
+        <differenceType>7012</differenceType>
+        <className>com/google/cloud/bigtable/beam/hbasesnapshots/ImportJobFromHbaseSnapshot</className>
+        <method>get*</method>
+        <method>set*</method>
```
Comment (vermas2012) [inline clirr-ignored-differences.xml:8]: Why is it a problem now and was OK before?

**Review by vermas2012 — COMMENTED** [review]
Overall looks good, minor comments

Lets validate that "--enableSnappy" is ok with Igor

File: bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/hbasesnapshots/ImportJobFromHbaseSnapshot.java [hunk 5]
```diff
@@ -79,6 +84,12 @@ public interface ImportOptions extends ImportJob.ImportOptions {
 
     @SuppressWarnings("unused")
     void setSnapshotName(String snapshotName);
+
+    @Description("Is importing Snappy compressed snapshot.")
+    Boolean getEnableSnappy();
```
Comment (igorbernstein2) [inline bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/hbasesnapshots/ImportJobFromHbaseSnapshot.java:89]:
```
@Default.Boolean(true)
boolean getEnableSnappy()
```

File: bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/hbasesnapshots/ImportJobFromHbaseSnapshot.java [hunk 6]
```diff
@@ -100,6 +111,19 @@ public static void main(String[] args) throws Exception {
   @VisibleForTesting
   static Pipeline buildPipeline(ImportOptions opts) throws Exception {
 
+    if(opts.getEnableSnappy() != null && opts.getEnableSnappy()) {
```
Comment (igorbernstein2) [inline bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/hbasesnapshots/ImportJobFromHbaseSnapshot.java:114]: use a primitive boolean with a default so you dont have to worry about null

File: bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/hbasesnapshots/ImportJobFromHbaseSnapshot.java [hunk 7]
```diff
@@ -100,6 +111,19 @@ public static void main(String[] args) throws Exception {
...
+        dataFlowOpts.setSdkContainerImage(CONTAINER_IMAGE_PATH_PREFIX + ReleaseInfo.getReleaseInfo().getVersion());
+        List<String> expOpts = dataFlowOpts.getExperiments();
+        if (expOpts != null) {
+            expOpts.add("use_runner_v2");
+            dataFlowOpts.setExperiments(expOpts);
+        } else {
+            final List<String> exps = new ArrayList<String>(Arrays.asList("use_runner_v2"));
+            dataFlowOpts.setExperiments(exps);
+        }
+    }
```
Comment (igorbernstein2) [inline bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/hbasesnapshots/ImportJobFromHbaseSnapshot.java:125]: You should avoid mutating lists inplace. Also, you should not add the flag if its already there:

```java
List<String> expOpts = Objects.firstNonNull(dataFlowOpts.getExperiments(), new ArrayList());
if (!expOpts.contains("use_runner_v2")) {
  expOpts = new ArrayList<>(existingOpts);
  expOpts.add("use_runner_v2")
}
dataflowOpts.setExperiments(expOpts);
```

File: bigtable-dataflow-parent/bigtable-beam-import/src/test/java/com/google/cloud/bigtable/beam/hbasesnapshots/EndToEndIT.java [hunk 8]
```diff
@@ -345,7 +346,7 @@ public void testHBaseSnapshotImport() throws Exception {
 
     // Validate the counters.
     Map<String, Long> counters = getCountMap(result);
-    Assert.assertEquals(counters.get("ranges_matched"), (Long) 101L);
+    Assert.assertEquals(counters.get("ranges_matched"), (Long) 100L);
```
Comment (igorbernstein2) [inline bigtable-dataflow-parent/bigtable-beam-import/src/test/java/com/google/cloud/bigtable/beam/hbasesnapshots/EndToEndIT.java:349]: should you add a different snapshot file for snappy?

File: bigtable-dataflow-parent/bigtable-beam-import/src/test/java/com/google/cloud/bigtable/beam/hbasesnapshots/EndToEndIT.java [hunk 9]
```diff
@@ -408,7 +409,57 @@ public void testHBaseSnapshotImportWithCorruptions() throws Exception {
...
+      if (items != null) {
+        allObjects.addAll(items);
+      }
+      nextToken = objects.getNextPageToken();
+    } while (nextToken != null);
+
+    List<StorageObject> myObjects =
+        allObjects.stream()
+            .filter(o -> o.getName().contains(importOpts.getJobName()))
+            .collect(Collectors.toList());
```
Comment (igorbernstein2) [inline bigtable-dataflow-parent/bigtable-beam-import/src/test/java/com/google/cloud/bigtable/beam/hbasesnapshots/EndToEndIT.java:447]: Now that the source snapshot is unique, you dont have to worry about filtering by job name. I think you can just check if there is anything in the restore directory (and dont have to pagniate)

File: clirr-ignored-differences.xml [hunk 10]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!-- see http://www.mojohaus.org/clirr-maven-plugin/examples/ignored-differences.html -->
+<differences>
+    <difference>
+        <differenceType>7012</differenceType>
+        <className>com/google/cloud/bigtable/beam/hbasesnapshots/ImportJobFromHbaseSnapshot</className>
```
Comment (igorbernstein2) [inline clirr-ignored-differences.xml:6]: should this be com/google/cloud/bigtable/beam/hbasesnapshots/ImportJobFromHbaseSnapsho/ImportOptions?

**Review by vermas2012 — CHANGES_REQUESTED** [review]
Please update the data generation script. https://paste.googleplex.com/5670577017716736

The generated zip file is at <>

File: bigtable-dataflow-parent/bigtable-beam-import/src/test/java/com/google/cloud/bigtable/beam/hbasesnapshots/EndToEndIT.java (hunk 9 above)
Comment (TracyCuiCan) [reply bigtable-dataflow-parent/bigtable-beam-import/src/test/java/com/google/cloud/bigtable/beam/hbasesnapshots/EndToEndIT.java]: Thanks!

File: bigtable-dataflow-parent/bigtable-beam-import/src/test/java/com/google/cloud/bigtable/beam/hbasesnapshots/EndToEndIT.java (hunk 8 above)
Comment (TracyCuiCan) [reply bigtable-dataflow-parent/bigtable-beam-import/src/test/java/com/google/cloud/bigtable/beam/hbasesnapshots/EndToEndIT.java:344]: snappy snapshot is generated to have the same data as normal snapshot.

File: bigtable-dataflow-parent/bigtable-beam-import/src/test/java/com/google/cloud/bigtable/beam/hbasesnapshots/EndToEndIT.java [hunk 11]
```diff
@@ -408,7 +409,57 @@ public void testHBaseSnapshotImportWithCorruptions() throws Exception {
 
     // Assert that the output collection is the right one.
     Map<String, Long> counters = getCountMap(result);
-    Assert.assertEquals(counters.get("ranges_matched"), (Long) 97L);
+    Assert.assertEquals(counters.get("ranges_matched"), (Long) 96L);
```
[the remark this answers is not in the document]
Comment (TracyCuiCan) [reply bigtable-dataflow-parent/bigtable-beam-import/src/test/java/com/google/cloud/bigtable/beam/hbasesnapshots/EndToEndIT.java:407]: Thanks!

**Review by vermas2012 — APPROVED** [review]

File: bigtable-dataflow-parent/bigtable-beam-import/src/test/resources/com/google/cloud/bigtable/beam/hbasesnapshots/EndToEndIT-snapshot-generate.sh [hunk 12]
```diff
@@ -1,126 +1,245 @@
...
-
-hbase shell <<EOF
-disable 'test'
-drop 'test'
-EOF
-
+hbase org.apache.hadoop.hbase.snapshot.ExportSnapshot -snapshot test-snappy-snapshot -copy-to /tmp/EndToEndIT-snapshot/data -mappers 16
+ 
+# Copy the data to local filesystem
+hadoop fs -copyToLocal /tmp/EndToEndIT-snapshot/ /tmp/
```
Comment (vermas2012) [inline bigtable-dataflow-parent/bigtable-beam-import/src/test/resources/com/google/cloud/bigtable/beam/hbasesnapshots/EndToEndIT-snapshot-generate.sh:226]: Move this to line 230.

File: bigtable-dataflow-parent/bigtable-beam-import/src/test/resources/com/google/cloud/bigtable/beam/hbasesnapshots/EndToEndIT-snapshot-generate.sh [hunk 13]
```diff
@@ -1,126 +1,245 @@
...
+hbase shell <<EOF
+disable 'test-table'
+drop 'test-table'
+ 
+disable 'test-snappy-table'
+drop 'test-snappy-table'
+ 
+delete_snapshot 'test-snapshot'
+delete_snapshot 'test-snappy-snapshot'
+EOF
```
Comment (vermas2012) [inline bigtable-dataflow-parent/bigtable-beam-import/src/test/resources/com/google/cloud/bigtable/beam/hbasesnapshots/EndToEndIT-snapshot-generate.sh:245]: Add the line 

```
# Cleanup the snapshot from HDFS 
hadoop fs -rm -r /tmp/EndToEndIT-snapshot
```

File: bigtable-dataflow-parent/bigtable-beam-import/src/test/resources/com/google/cloud/bigtable/beam/hbasesnapshots/EndToEndIT-snapshot-generate.sh (hunk 12 above)
Comment (TracyCuiCan) [reply bigtable-dataflow-parent/bigtable-beam-import/src/test/resources/com/google/cloud/bigtable/beam/hbasesnapshots/EndToEndIT-snapshot-generate.sh]: Done, thanks!

File: bigtable-dataflow-parent/bigtable-beam-import/src/test/resources/com/google/cloud/bigtable/beam/hbasesnapshots/EndToEndIT-snapshot-generate.sh [hunk 14]
```diff
@@ -242,4 +242,6 @@ drop 'test-snappy-table'
  
 delete_snapshot 'test-snapshot'
 delete_snapshot 'test-snappy-snapshot'
-EOF
\ No newline at end of file
+EOF
+
+hadoop fs -rm -r /tmp/EndToEndIT-snapshot
```
Comment (vermas2012) [inline bigtable-dataflow-parent/bigtable-beam-import/src/test/resources/com/google/cloud/bigtable/beam/hbasesnapshots/EndToEndIT-snapshot-generate.sh:247]: Please add the comment  as well.

File: bigtable-dataflow-parent/bigtable-beam-import/src/test/java/com/google/cloud/bigtable/beam/hbasesnapshots/EndToEndIT.java [hunk 15]
```diff
@@ -403,7 +404,48 @@ public void testHBaseSnapshotImportWithCorruptions() throws Exception {
...
+    String restorePathPrefix =
+        CleanupHBaseSnapshotRestoreFilesFn.getListPrefix(
+            HBaseSnapshotInputConfigBuilder.RESTORE_DIR);
+    List<StorageObject> allObjects = new ArrayList<>();
+    Objects objects = gcsUtil.listObjects(bucket, restorePathPrefix, null);
+    List<StorageObject> items = objects.getItems();
+    if (items != null) {
+      allObjects.addAll(items);
+    }
+    Assert.assertTrue("Restore directory wasn't deleted", allObjects.isEmpty());
```
Comment (igorbernstein2) [inline bigtable-dataflow-parent/bigtable-beam-import/src/test/java/com/google/cloud/bigtable/beam/hbasesnapshots/EndToEndIT.java:434]: I think allObjects is relic from the previous impl. Please cleanup:
```
Objects restoreFiles = gcsUtil.listObjects(bucket, restorePathPrefix, null);
Assert.assertTrue("Restore directory wasn't deleted", restoreFiles.isEmpty());
```

**Pull request merged**